File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    
    const stack = []; 

    const compare = (target,original) => {
        if(original === "(") return target === ")";
        if(original === "[") return target === "]";
        if(original === "{") return target === "}";
        return false;
    }
    const maxLength = s.length;
    for(let i = 0 ; i < maxLength;i++ ){
        if(stack.size !== 0 && compare(s[i], stack[stack.length-1])){
            stack.pop();
            continue; 
        } 
        stack.push(s[i]);
    }
    return stack.length === 0 
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        var stack = new ArrayDeque<>();

        var map = Map.of(
            ')', '(', 
            '}', '{', 
            ']', '['
            );

        for(var c: s.toCharArray()){
            if(map.containsKey(c)) {
                if(stack.isEmpty() || stack.peek() != map.get(c)) return false;
                stack.pop();
            }else {
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }
}